src/queue/multilevel.rs

```diff
@@ -0,0 +1,835 @@
+            return;
+        }
+        // The statistics may be not so accurate because we cannot load two
+        // atomics at the same time.
+        let total = self.total_elapsed.0.load(SeqCst);
+        if Duration::from_micros(total) < ADJUST_CHANCE_INTERVAL {
+            // Another thread just adjusted the chances.
+            return;
+        }
+        let total = self.total_elapsed.0.swap(0, SeqCst);
+        let level0 = self.level0_elapsed.0.swap(0, SeqCst);
+        let current_proportion = level0 as f64 / total as f64;
+        let diff = self.level0_proportion_target - current_proportion;
+        let level0_chance = self.level0_chance.load(SeqCst);
+        let new_chance = if diff > ADJUST_CHANCE_THRESHOLD {
+            cmp::min(
+                level0_chance.saturating_add(ADJUST_AMOUNT),
+                MAX_LEVEL0_CHANCE,
+            )
+        } else if diff < -ADJUST_CHANCE_THRESHOLD {
+            cmp::max(level0_chance - ADJUST_AMOUNT, MIN_LEVEL0_CHANCE)
+        } else {
+            level0_chance
+        };
+        self.level0_chance.store(new_chance, SeqCst);
+        self.adjusting.store(false, SeqCst);
+    }
+}
+
+#[derive(Default, Debug)]
+pub(crate) struct ElapsedTime(AtomicU64);
+
+impl ElapsedTime {
+    fn as_duration(&self) -> Duration {
+        Duration::from_micros(self.0.load(SeqCst))
+    }
+
+    fn inc_by(&self, t: Duration) -> u64 {
+        self.0.fetch_add(t.as_micros() as u64, SeqCst)
+    }
+
+    #[cfg(test)]
+    fn from_duration(dur: Duration) -> Self {
+        ElapsedTime(AtomicU64::new(dur.as_micros() as u64))
+    }
+}
+
+thread_local!(static TLS_LAST_CLEANUP_TIME: Cell<Instant> = Cell::new(Instant::now()));
+
+struct TaskElapsedMap {
+    new_index: AtomicUsize,
+    maps: [DashMap<u64, Arc<ElapsedTime>>; 2],
+    cleanup_interval: Duration,
+    last_cleanup_time: RwLock<Instant>,
+    cleaning_up: AtomicBool,
+}
+
+impl Default for TaskElapsedMap {
+    fn default() -> TaskElapsedMap {
+        TaskElapsedMap::new(DEFAULT_CLEANUP_OLD_MAP_INTERVAL)
+    }
+}
+
+impl TaskElapsedMap {
+    fn new(cleanup_interval: Duration) -> TaskElapsedMap {
+        TaskElapsedMap {
+            new_index: AtomicUsize::new(0),
+            maps: Default::default(),
+            cleanup_interval,
+            last_cleanup_time: RwLock::new(now()),
+            cleaning_up: AtomicBool::new(false),
+        }
+    }
+
+    fn get_elapsed(&self, key: u64) -> Arc<ElapsedTime> {
+        let new_index = self.new_index.load(SeqCst);
+        let new_map = &self.maps[new_index];
+        let old_map = &self.maps[new_index ^ 1];
+        if let Some(v) = new_map.get(&key) {
+            return v.clone();
+        }
+        let elapsed = match old_map.get(&key) {
+            Some(v) => {
+                new_map.insert(key, v.clone());
+                v.clone()
+            }
+            _ => {
+                let v = new_map.get_or_insert_with(&key, Default::default);
```

Comment:
What if `new_index` is changed during two concurrent `get_elapsed`?

Reply:
The two elapsed can be different. It can make the statistics inaccurate but I can't find out a good way to solve it now.